capstone-rs/src/instruction.rs

```diff
@@ -214,53 +171,153 @@ impl<'a> Insn<'a> {
+        InsnId(self.insn().id)
     }
 
     /// Size of instruction (in bytes)
-    fn len(&self) -> usize {
-        self.insn.size as usize
+    pub fn len(&self) -> usize {
+        self.insn().size as usize
     }
 
     /// Instruction address
     pub fn address(&self) -> u64 {
-        self.insn.address as u64
+        self.insn().address as u64
     }
 
     /// Byte-level representation of the instruction
     pub fn bytes(&self) -> &[u8] {
-        &self.insn.bytes[..self.len()]
+        &self.insn().bytes[..self.len()]
     }
 
     /// Returns the `Detail` object, if there is one. It is up to the caller to determine
     /// the pre-conditions are satisfied.
     ///
     /// Be careful this is still in early stages and largely untested with various `cs_option` and
     /// architecture matrices
+    ///
+    /// # Safety
+    /// The [`cs_insn::detail`] pointer must be valid and non-null.
     pub(crate) unsafe fn detail(&self, arch: Arch) -> InsnDetail {
-        InsnDetail(&*self.insn.detail, arch)
+        InsnDetail(&*self.insn().detail, arch)
+    }
+
+    // Any pointers are owned by the Capstone C library and should be freed accordingly
+    #[inline(always)]
+    fn insn(&self) -> &cs_insn {
+        &self.insn
     }
 }
 
+/// A single disassembled CPU instruction.
+///
+/// # Detail
+///
+/// To learn how to get more instruction details, see [`InsnDetail`].
+#[repr(transparent)]
+pub struct Insn<'a> {
+    /// Inner `cs_insn`
+    pub(crate) insn: cs_insn,
+
+    /// Adds lifetime
+    pub(crate) _marker: PhantomData<&'a InsnDetail<'a>>,
+}
+
+impl<'a> From<&Insn<'_>> for OwnedInsn<'a> {
+    // SAFETY: assumes that `cs_detail` struct transitively only contains owned
+    // types and no pointers, including the union over the architecture-specific
+    // types.
+    fn from(insn: &Insn<'_>) -> Self {
+        let mut new = unsafe { <*const cs_insn>::read(&insn.insn as _) };
+        new.detail = if new.detail.is_null() {
+            new.detail
+        } else {
+            unsafe {
+                let new_detail = Box::new(*new.detail);
+                Box::into_raw(new_detail)
+            }
+        };
+        Self {
+            insn: new,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// SAFETY:
+/// 1. `Insn` must be `#repr(transparent)` of `cs_insn`
+/// 2. all `Insn` methods must be safe to perform for an `OwnedInsn`
+impl<'a> Deref for OwnedInsn<'a> {
+    type Target = Insn<'a>;
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*(&self.insn as *const cs_insn as *const Insn) }
+    }
+}
+
+/// A single disassembled CPI instruction that lives on the Rust heap.
+///
+/// # Detail
+///
+/// To learn how to get more instruction details, see [`InsnDetail`].
+pub struct OwnedInsn<'a> {
+    /// Inner cs_insn
+    pub(crate) insn: cs_insn,
+
+    /// Adds lifetime
+    pub(crate) _marker: PhantomData<&'a InsnDetail<'a>>,
+}
+
+/// Contains architecture-independent details about an [`Insn`].
```

Comment:
could you hoist this back up higher in the file (where it was before) to simplify this change?